advisories/a64325.md

```diff
@@ -0,0 +1,41 @@
+---
+title: Technical Advisory 64325
+summary: Race condition between long running reads and replica removal
+toc: true
+---
+
+Publication date: May 3, 2021
+
+## Description
+
+Cockroach Labs has discovered a race condition, where a long running read request submitted during replica removal may be evaluated on the removed replica, returning an empty result. Common causes of replica removal are range merges and replica rebalancing in the cluster. When a replica is removed there is a small window, where the data is already deleted but a read request will still see it as a valid replica. In this case a read request will always return an empty result for the part of the query that relied on this range.
+
+This is a very rare scenario that can only be encountered, if a number of very unlikely conditions are met:
+
+ - A read request checks that the node is holding the lease and begins processing
+ - The node loses the lease before the read request is able to lock the replica for reading
+ - The read request checks the state of the replica and makes sure it is alive
+ - The new leaseholder decides to immediately removes the old leaseholder replica from the range, due to a rebalance or range merge.
+ - The removal is processed before the read requests starts reading the data from the replica
+ - The read request observes the replica in a post-deletion state returning an empty result set.
+
+ Due to the nature of this race condition, we believe that it is only likely for this to happen on an extremely overloaded node that is struggling to keep up with processing requests and experiences natural delays between each step in the processing of a read request.
+
+This issue affects CockroachDB [versions](/docs/releases/) 19.2 and later.
```

Comment:
Backported a test for this to 19.2 (#64510), and as far as I can tell it does not have this bug -- it appears to have been introduced in 20.1.

Reply:
Thanks for doing that